src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -81,7 +82,14 @@ public class RSessionProvider : IRSessionProvider {
                 return;
             }
 
-            foreach (var session in _sessions.Values) {
+            var sessions = GetSessions();
+            var stopHostTasks = sessions.Select(session => session.StopHostAsync());
+            try {
+                Task.WhenAll(stopHostTasks).GetAwaiter().GetResult();
```

Comment:
MEF can call `Dispose` from any thread, but you're probably right that VS should wait until its MEF components are disposed.